### PR TITLE
[WIP] Support for tracking Exception Sets with ValueNumbers

### LINKED ISSUE #8648: [RyuJIT] Expression a % 0 sometimes generates incorrect code.
The following example:
```C#
[MethodImpl(MethodImplOptions.NoInlining)]
static int Test(sbyte x)
{
    return x % 0;
}
```
generates
```asm
G_M55886_IG02:
       480FBEC1             movsx    rax, cl
G_M55886_IG03:
       C3                   ret
```
Of course, this doesn't generate a `DivideByZeroException` exception as it should.

This is partially caused by a recent change I did that transforms `x % c` into `x - (x / c) * c` if c is not a power of 2. 0 is not a power of 2 but the transform should not be done when c is 0. ARM64 appears to do the same transform unconditionally so probably it is also affected by this. I think we should simply transform `x % 0` into `x / 0`.

However, the actual bug is somewhere else. The following IL code triggers it as well:
```
.method private hidebysig static int32 
        Test(int8 arg) cil managed noinlining
{
        .maxstack  8
        ldarg.0
        ldarg.0
        ldc.i4.0
        div
        ldc.i4.0
        mul
        sub
        ret
}
```
The C# compiler doesn't usually produce this kind of code, it drops `x * 0`. It's that multiplication that's probably mishandled by the JIT, it tries to drop it and sometimes it drops the division as well. When it works correctly the generated code looks like this:
```asm
G_M55886_IG02:
       480FBEC9             movsx    rcx, cl
       4533C0               xor      r8d, r8d
       8BC1                 mov      eax, ecx
       99                   cdq
       41F7F8               idiv     edx:eax, r8d
       8BC1                 mov      eax, ecx
```

I'll look into this. 
category:correctness
theme:cse
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by briansull**
Added support to create and track Exception sets with ValueNumbers
Some cleanup in ValueNumbering as well.
The Asm Diffs produced are  fairly minimal with this set of changes.
Fixes #8648

### CODE REVIEW

**Comment by briansull** [comment]
@mikedn @AndyAyersMS  PTAL

I will be on vacation some/most of next week.

File: src/jit/valuenum.cpp [hunk 1]
```diff
@@ -155,7 +155,7 @@ T ValueNumStore::EvalOp(VNFunc vnf, T v0, T v1, ValueNum* pExcSet)
             case GT_DIV:
                 if (IsIntZero(v1))
                 {
-                    *pExcSet = VNExcSetSingleton(VNForFunc(TYP_REF, VNF_DivideByZeroExc));
+                    *pExcSet = VNExcSetSingleton(VNForFunc(TYP_REF, VNF_DivideByZeroExc, VNForVoid()));
```
Comment (mikedn) [inline src/jit/valuenum.cpp]: Is VNForVoid correct here? The description of VNF_DivideByZeroExc is the divisor value so it seems that here it should the VN of constant 0. Same comment for the next 2 changes bellow.

File: src/jit/assertionprop.cpp [hunk 2]
```diff
@@ -1434,12 +1435,19 @@ bool Compiler::optIsTreeKnownIntValue(bool vnBased, GenTree* tree, ssize_t* pCon
...
-    if (!vnStore->IsVNConstant(tree->gtVNPair.GetConservative()))
+    ValueNum vn = vnStore->VNNormVal(tree->gtVNPair.GetConservative());
+    if (!vnStore->IsVNConstant(vn))
+    {
+        return false;
+    }
+    // ValueNumber indicates that this node evaluates to a constant
+
+    // But we could have an integer division by zero
+    if (tree->OperMayThrow(this))
```
Comment (mikedn) [inline src/jit/assertionprop.cpp]: Hmm, why is this needed?

File: src/jit/assertionprop.cpp [hunk 3]
```diff
@@ -1764,14 +1772,15 @@ AssertionInfo Compiler::optCreateJTrueBoundsAssertion(GenTree* tree)
     GenTree* op1 = relop->gtGetOp1();
     GenTree* op2 = relop->gtGetOp2();
 
-    ValueNum vn = op1->gtVNPair.GetConservative();
+    ValueNum vn      = vnStore->VNNormVal(op1->gtVNPair.GetConservative());
+    ValueNum relopVN = vnStore->VNNormVal(relop->gtVNPair.GetConservative());
+    ValueNum op2VN   = vnStore->VNNormVal(relop->gtVNPair.GetConservative());
```
Comment (mikedn) [inline src/jit/assertionprop.cpp]: relop should be op2

File: src/jit/assertionprop.cpp [hunk 4]
```diff
@@ -1764,14 +1772,15 @@ AssertionInfo Compiler::optCreateJTrueBoundsAssertion(GenTree* tree)
     GenTree* op1 = relop->gtGetOp1();
     GenTree* op2 = relop->gtGetOp2();
 
-    ValueNum vn = op1->gtVNPair.GetConservative();
+    ValueNum vn      = vnStore->VNNormVal(op1->gtVNPair.GetConservative());
```
Comment (mikedn) [inline src/jit/assertionprop.cpp]: Might be nice to rename to op1VN

File: src/jit/assertionprop.cpp [hunk 5]
```diff
@@ -2373,6 +2381,15 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* stmt, Gen
     ValueNum vnCns = tree->gtVNPair.GetConservative();
     ValueNum vnLib = tree->gtVNPair.GetLiberal();
 
+    // We typically want to use the Normal ValueNumber when checking for constants
+    // except for the nodes that might generate an exception like GT_DIV, GT_MOD, etc...
+    //
+    if (!tree->OperMayThrow(this))
```
Comment (mikedn) [inline src/jit/assertionprop.cpp]: So I guess this is needed because if we have a VN exception set we don't really know if it comes from the node itself or from its children?

File: src/jit/assertionprop.cpp [hunk 6]
```diff
@@ -3058,9 +3075,9 @@ GenTree* Compiler::optAssertionPropGlobal_RelOp(ASSERT_VALARG_TP assertions, Gen
     bool allowReverse = true;
 
     // If the assertion involves "op2" and it is a constant, then check if "op1" also has a constant value.
-    if (vnStore->IsVNConstant(op2->gtVNPair.GetConservative()))
+    ValueNum vnCns = vnStore->VNNormVal(op2->gtVNPair.GetConservative());
```
Comment (mikedn) [inline src/jit/assertionprop.cpp]: There are so many of these VNNormVal that perhaps we should add a shortcut to `Compiler` - `ValueNum Compiler::vnConservativeNormVal(GenTree*);`

File: src/jit/optcse.cpp [hunk 7]
```diff
@@ -651,11 +651,62 @@ unsigned Compiler::optValnumCSE_Index(GenTree* tree, GenTree* stmt)
...
+    // (alternatively we could use a subset operation on the exc sets)
+    //
+    // One exception to using the normal value is for the GT_COMMA nodes.
+    // So we check to see if we have a GT_COMMA with a different value number
+    // than the one from its op2.  For this case we create two CSE candidates.
+    // This allows us to CSE the GT_COMMA separately from its value.
+    //
+    ValueNum vnLibNorm = vnStore->VNNormVal(vnLib);
+
+    // We assign either vnLib or vnLinNorm as the hash key
```
Comment (mikedn) [inline src/jit/optcse.cpp]: vnLinNorm -> vnLibNorm

File: src/jit/optcse.cpp [hunk 8]
```diff
@@ -651,11 +651,62 @@ unsigned Compiler::optValnumCSE_Index(GenTree* tree, GenTree* stmt)
...
+        {
+            key = (unsigned)vnLibNorm;
+        }
 
-    /* Compute the hash value for the expression */
+        // If we didn't do the above we would have op1 as the CSE def
+        // and the parent comma as the CSE use (but with a different exc set)
+        // This would prevent us from making any CSE with the comma
+        //
+        assert(vnLibNorm == vnStore->VNNormVal(vnOp2Lib)); // (We don't need to assert this)
```
Comment (mikedn) [inline src/jit/optcse.cpp]: The comment after the assert seems strange

File: src/jit/assertionprop.cpp (hunk 6 above)
Comment (briansull) [reply src/jit/assertionprop.cpp]: Yes, I am considering adding  heper functions:

vnConservativeNormVal
vnLiberalNormVal

**Comment by BruceForstall** [comment]
@briansull Is this still WIP? Is it ready to be reviewed, or are you still working on it? (btw, it has merge conflicts now)

**Comment by briansull** [comment]
I decided to stage these changes by checking this stuff using an series of smaller chnage sets.
Currently I have this part our for review: #19720

**Comment by briansull** [comment]
Checked in with #20129
